// tsun/bin/src/executor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var path = require("path");
var child_process = require("child_process");
var fs = require("fs");
function runCode(argv) {
    // run code in temp path, and cleanup
    var temp = require('temp');
    temp.track();
    process.on('SIGINT', function () { return temp.cleanupSync(); });
    process.on('SIGTERM', function () { return temp.cleanupSync(); });
    var tempPath = temp.mkdirSync('tsrun');
    var outDir = tempPath;
    if (argv.o) {
        outDir = path.join(tempPath, argv.o);
    }
    var compileError = compile(argv._, {
        outDir: outDir,
        noEmitOnError: true,
        target: ts.ScriptTarget.ES5,
        module: ts.ModuleKind.CommonJS,
        experimentalDecorators: true,
    });
    if (compileError)
        process.exit(compileError);
    linkDir(process.cwd(), tempPath);
    // slice argv. 0: node, 1: tsun binary 2: arg
    var newArgv = process.argv.slice(2).map(function (arg) {
        if (!/\.ts$/.test(arg))
            return arg;
        return path.join(outDir, arg.replace(/ts$/, 'js'));
    });
    child_process.execFileSync('node', newArgv, {
        stdio: 'inherit'
    });
    process.exit();
}
exports.runCode = runCode;
function linkDir(src, dest) {
    var files = ['node_modules', 'typings'];
    for (var _i = 0, files_1 = files; _i < files_1.length; _i++) {
        var file = files_1[_i];
        var srcpath = path.join(src, file);
        var destpath = path.join(dest, file);
        fs.symlinkSync(srcpath, destpath, 'dir');
    }
}
function compile(fileNames, options) {
    var program = ts.createProgram(fileNames, options);
    var emitResult = program.emit();
    var allDiagnostics = ts.getPreEmitDiagnostics(program).concat(emitResult.diagnostics);
    allDiagnostics.forEach(function (diagnostic) {
        var message = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
        if (!diagnostic.file)
            return console.log(message);
        var _a = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start), line = _a.line, character = _a.character;
        console.log(diagnostic.file.fileName + " (" + (line + 1) + "," + (character + 1) + "): " + message);
    });
    var exitCode = emitResult.emitSkipped ? 1 : 0;
    return exitCode;
}
